cpp/src/io/utilities/datasource.cpp

```diff
@@ -54,16 +54,16 @@ class arrow_io_source : public datasource {
 
   size_t host_read(size_t offset, size_t size, uint8_t *dst) override
   {
-    int64_t bytes_out = 0;
-    CUDF_EXPECTS(arrow_file->ReadAt(offset, size, &bytes_out, dst).ok(), "Cannot read file data");
-    return bytes_out;
+    arrow::Result<int64_t> result = arrow_file->ReadAt(offset, size, dst);
```

Comment:
Is there a reason why `arrow::Result<int64_t>` is explicit here, but `size` below is `auto`? Minor nitpick, but it looks inconsistent. `result` is a signed `int64_t` but it's returned as unsigned `size_t`?

Also, I love this Result API; it's very Rust-like.

Reply:
> Is there a reason why arrow::Result<int64_t> is explicit here, but size below is auto?

Yeah I'd originally used auto, but changed it to `Result<int64_t>` once I saw the diff. This call is replacing the `int64_t bytes_out` declaration, and the explicit type makes that more obvious in the diff.

> Minor nitpick, but it looks inconsistent. result is a signed int64_t but it's returned as unsigned size_t?

Yeah that doesn't seem right, but changing the return type might affect other places. I have a hunch Arrow's chosen to return `int64_t` because it's platform independent, not because they're actually returning -1? At least, I hope so.

Reply:
So... I changed it back to auto :sweat_smile:.

Reply:
At least they're _all_ `auto` now ¯\\\_(ツ)_/¯
If you're sure the `int64_t` here is always positive, then I guess it's fine to have an unchecked cast to `size_t`.